File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/2512-예산/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 10005

int N, M;
int A[NM];

void input() {
	cin >> N;
	for (int i = 1; i <= N; i++) cin >> A[i];
	cin >> M;
}

bool determination(int limit) {
	int sum = 0;
	for (int i = 1; i <= N; i++) {
		sum += min(A[i], limit);
	}
	return sum <= M;
}

void pro() {
	int L = 0, R = 0, ans = 0;
	for (int i=1;i<=N;i++) R = max(R, A[i]);
	while (L <= R) {
		int mid = (L + R) / 2;
		if (determination(mid)) {
			ans = mid;
			L = mid + 1;
		}
		else R = mid - 1;
	}
	cout << ans;
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}

File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/2512-예산/solution.py
import sys
si = sys.stdin.readline
n = int(si())
a = list(map(int, si().split()))
m = int(si())

def determination(limit):
    sum = 0
    for x in a:
        sum += min(x, limit)
    return sum <= m

l, r, ans = 0, max(a), 0
while l <= r:
    mid = (l + r) // 2
    if determination(mid):
        ans = mid
        l = mid + 1
    else:
        r = mid - 1

print(ans)